File: Module1 - Zachary Abbe -CS499/Project1-ZacharyAbbe-CS320/ContactService.java
/********************
 * Name: Zachary Abbe
 * Course: CS-320
 * Date: 04/12/2025
 * Desc: This contactService class contains the necessary capabilities of creating a list of contacts, while adding new ones and deleting them, as well as updates.
 ********************/
import java.util.ArrayList;

public class ContactService {

    // List to store contacts
    private final ArrayList<Contact> contacts = new ArrayList<>();

    // Adds a new contact if the ID is unique
    public boolean addContact(Contact contact) {
        for (Contact c : contacts) {
            if (c.getId().equalsIgnoreCase(contact.getId())) {
                throw new IllegalArgumentException("Duplicate ID");
            }
        }
        contacts.add(contact);
        return true;
    }

    // Deletes a contact by ID
    public void deleteContact(String id) {
        contacts.removeIf(c -> c.getId().equalsIgnoreCase(id));
    }

    // Updates the first name of a contact by ID
    public void updateFirstName(String id, String firstName) {
        Contact contact = findContactById(id);
        if (contact != null) {
            contact.setFirstName(firstName);
        } else {
            throw new IllegalArgumentException("Contact not found");
        }
    }

    // Updates the last name of a contact by ID
    public void updateLastName(String id, String lastName) {
        Contact contact = findContactById(id);
        if (contact != null) {
            contact.setLastName(lastName);
        } else {
            throw new IllegalArgumentException("Contact not found");
        }
    }

    // Updates the phone number of a contact by ID
    public void updatePhoneNumber(String id, String phoneNumber) {
        Contact contact = findContactById(id);
        if (contact != null) {
            contact.setPhoneNumber(phoneNumber);
        } else {
            throw new IllegalArgumentException("Contact not found");
        }
    }

    // Updates the address of a contact by ID
    public void updateAddress(String id, String address) {
        Contact contact = findContactById(id);
        if (contact != null) {
            contact.setAddress(address);
        } else {
            throw new IllegalArgumentException("Contact not found");
        }
    }

    // Helper method to find a contact by ID
    private Contact findContactById(String id) {
        for (Contact c : contacts) {
            if (c.getId().equalsIgnoreCase(id)) {
                return c;
            }
        }
        return null;
    }

    // Getter for test access
    public ArrayList<Contact> getContacts() {
        return contacts;
    }
}

File: CS499 Project1 - ZacharyAbbe/ContactService.cpp
#include "ContactService.h"
#include <stdexcept>
#include <algorithm>

using namespace std;

bool ContactService::addContact(const Contact& contact)
{
	//for (auto& c : contacts)
	//{
	//	if (c.getId() == contact.getId())
	//	{
	//		throw invalid_argument("Duplicate ID");
	//	}
	//}
	//contacts.push_back(contact);
	//return true;
	if (contacts.count(contact.getId()))
	{
		throw invalid_argument("Duplicate ID");
	}
	contacts.insert({ contact.getId(), contact });
	return true;
}
void ContactService::deleteContact(const string& id)
{	// we won't need to use a complicated lambda functiion to delete a contact.
	//contacts.erase(remove_if(contacts.begin(), contacts.end(),[&](const Contact& c) { return c.getId() == id;}), contacts.end());
	contacts.erase(id);
}

Contact* ContactService::findContactById(const string& id)
{
	//for (auto& c : contacts)
	//{
	//	if (c.getId() == id)
	//	{
	//		return &c;
	//	}
	//}
	// we will use the maps method to find the contact
	auto it = contacts.find(id);
		if (it != contacts.end())
		{
			//it is the iterator and it->second gives us the value of the and we can return its address
			return &it->second;
		}
	return nullptr; //return null if not found
}

void ContactService::updateFirstName(const string& id, const string& firstName) 
{
	Contact* contact = findContactById(id);
	if (contact) 
		{
			contact->setFirstName(firstName);
		}
	else 
		{
			throw invalid_argument("Contact not found");
		}
}
void ContactService::updateLastName(const string& id, const string& lastName) 
{
	Contact* contact = findContactById(id);
	if (contact) 
		{
			contact->setLastName(lastName);
		}
	else 
		{	
			throw invalid_argument("Contact not found");
		}
}

void ContactService::updatePhoneNumber(const string& id, const string& phoneNumber) 
{
	Contact* contact = findContactById(id);
	if (contact) 
		{
			contact->setPhoneNumber(phoneNumber);
		}
	else 
		{
			throw invalid_argument("Contact not found");
		}
}

void ContactService::updateAddress(const string& id, const string& address)	
{
	Contact* contact = findContactById(id);
	if (contact) 
		{
			contact->setAddress(address);
		}
	else 
		{
			throw invalid_argument("Contact not found");
		}
}

vector<Contact> ContactService::getContacts() const {
	vector<Contact> contactList;
	for (const auto& pair : contacts) {
		contactList.push_back(pair.second);
	}
	return contactList;  //  return by value 
}